Front end build housekeeping

kategreenUSDS opened the pull request:
Lots of big little things:
* Removing `gulp` for a more straightforward css build process
* Upgrading `uswds` from 2.12.0 to 3.x.x. **Note**: this does not include upgrading the styles at all, just the underlying build process and the version of `uswds` it's building against
* Updating the `Dockerfile` to reflect the more streamlined build process without `gulp`
* Updating `package.json` to add a new `build-css` script and use it. Also removing a bunch of dependencies
* Adding versions to `package.json` and `Gemfile`

Running on https://website-staging-wise-cat-bk.app.cloud.gov

Addresses usds/website-management#103 and usds/website-management#92

kategreenUSDS committed:
remove gulp, add uswds3, update Dockerfile to reflect

kategreenUSDS committed:
new style compilation scss and compiled css

kategreenUSDS committed:
compress the generated css, add flag to only build css when there are
changes

kategreenUSDS committed:
major documentation upgrade to reflect docker and uswds changes

lisac-usds commented on README.md:
```diff
@@ -5,53 +5,59 @@ The USDS website is built with:
+If you want to make changes and have them update on the running container
+we need to map the active source code in to the container with a volume mount and
+expose the port for livereload. This should ensure that changes you make are being
+mapped to the /app director and refresh on the running container. Note the volume mount syntax
```
```suggestion
mapped to the /app directory and refresh on the running container. Note the volume mount syntax
```

SteveMattinglyUSDS replied:
`/app` director***y***

lisac-usds reviewed (commented):
This looks really well done - great work, @kategreenUSDS.
Confirming the docker local dev and README instructions worked for me.

A minor thing: the caret on the top banner (official govt website, here's how you know) renders differently from what i'm seeing in the production environment. Is that something that could be tracked down before merging this in?  from my local dev:
- chrome: shows as a  blue square <img src="https://user-images.githubusercontent.com/31673962/211074582-dd987fd3-2995-41b8-aabc-a36fec51ac2c.png" width='200'>
- firefox: nothing visible to me <img src="https://user-images.githubusercontent.com/31673962/211074563-b315043a-cfcd-4af5-bade-0938cfe1b874.png" width='200'>     

Great work!

kategreenUSDS committed:
bring up to USWDS3.0 baseline for custom scss. there's still more
to do in here to simplify and stuff, but this is the baseline

kategreenUSDS commented:
@lisac-usds The caret thing is ... familiar to me. it's something to do with the chrome browser. let me search through my notes and find the solution we applied on the CTC/tax project.

kategreenUSDS committed:
adding autoprefixer

kategreenUSDS committed:
autoprefixer is js now
setting js and ruby version in package.json and Gemfile as well as Dockerfile
removing a few extra dependencies

kategreenUSDS commented:
Fixed the bug that Lisa found!

<img width="719" alt="Screenshot 2023-01-18 at 10 52 58 AM" src="https://user-images.githubusercontent.com/105316441/213221572-70ec05c3-a933-40ba-9a40-efec3713c5a3.png">

rachellanman commented on README.md:
````diff
@@ -83,50 +77,3 @@ Updating USWDS may overwrite custom settings, break the theme and manifest files
-for `pwd` might be different on Windows or a non Bash shell. 
-```
-docker run -p 4080:4000  -p 35729:35729 -v $(pwd):/app --name usdsweb usds-website
-```
````
nice cleanup of the readme! 🎉

rachellanman commented on assets/stylesheets/application/_variables.scss:
```diff
@@ -1,62 +1,9 @@
 // USWDS Overrides
```
this file now just contains comments - does it need to exist?

kategreenUSDS replied:
good question. yes, i anticipate using this as i further consolidate the css

rachellanman commented on assets/stylesheets/uswds/styles.scss:
```diff
@@ -86,7 +86,7 @@
 // Import USDS custom styles
 
-@import "../application/variables";
+//@import "../application/variables";
```
can this just be deleted, instead of commented out?

kategreenUSDS replied:
good catch

rachellanman commented on assets/stylesheets/uswds/uswds-settings.scss:
```diff
@@ -0,0 +1,49 @@
+$color-usds-blue-medium:          #064e96; // 508 compliant
```
nice compliance callout

rachellanman replied:
future us will thank you 😆

rachellanman reviewed (dismissed):
just a couple small things. Overall it looks great

BillChapmanUSDS commented:
One thing I noticed is that the Docker pulls from 2.7 and the Gemfile pegs 2.7.7 This only works because 2.7.7 is the latest ruby on version 2. But if a security patch comes out and bounces  that point release,  a clean build without a docker cache will pull 2.7.8  and then the bundle will fail. 

I think the best fix is to use loose versioning in the Gemfile since ruby 2.7 is no longer receiving feature releases the 2.7.x should remain consistent and only update for security patches.  

so if we do `ruby "~> 2.7.0"`  in the Gemfile  it should survive security patches being made to the base image. 

I'm not a fan of using flexible versioning beyond a point release though. 2.7.x is great but 2.x is trouble.  

However, and I don't think this is the case but if something else really needs to be specifically pegged to 2.7.7 we should make sure we pull that image specifically.

kategreenUSDS committed:
review feedback
* ruby version change in Gemfile
* commented out file in the root scss file

SteveMattinglyUSDS commented on README.md:
```diff
@@ -5,53 +5,59 @@ The USDS website is built with:
+- [Ruby](https://www.ruby-lang.org/en/documentation/installation/), best set up on version 2.7 for Jekyll
 - [NPM](https://github.com/npm/cli)
-- [Gulp CLI](https://gulpjs.com/docs/en/getting-started/quick-start)
+- Docker
```
Minor suggestion: this stands out as the only bullet that's not a link. Maybe link `https://www.docker.com/get-started/`    ?

SteveMattinglyUSDS commented on README.md:
```diff
@@ -5,53 +5,59 @@ The USDS website is built with:
+we need to map the active source code in to the container with a volume mount and
+expose the port for livereload. This should ensure that changes you make are being
+mapped to the /app director and refresh on the running container. Note the volume mount syntax
+for `pwd` might be different on Windows or a non Bash shell. 
```
Good to give a hint to non bash folks, but if they need the hint this probably doesn't give them what they need to solve it. Suggest something like

> In bash, `$(pwd)` inserts the path of the current working directory into the command shown. If that does not work for you, try replacing that portion of the command with the correct path.

SteveMattinglyUSDS reviewed (commented):
Sorry for the delayed review. I can't contribute much on SCSS stuff, but I marked one typo and some suggestions. One other problem that I couldn't comment because it wasn't touched in this PR: the last link under README's Change Theme Settings is broken (non-existent anchor).

I noticed two other things that I'll put here as questions.
1. Why do several of the package.json scripts use single ampersands between commands? (I know what it means, and apparently it works, but I'm interested in the context.)
2. This is a public repo with instructions that the public can't use (cloud.gov access, etc.) Does USDS have any practices for handling such situations?
